crates/solver/src/settlement/settlement_encoder.rs

```diff
@@ -260,7 +369,19 @@ impl SettlementEncoder {
     pub fn finish(mut self) -> EncodedSettlement {
         self.drop_unnecessary_tokens_and_prices();
 
-        let clearing_prices = self
+        let (mut liquidity_order_buy_tokens, mut liquidity_order_prices): (Vec<H160>, Vec<U256>) =
+            self.liquidity_order_trades
+                .iter()
+                .map(|liquidity_order_trade| {
+                    (
+                        liquidity_order_trade.trade.order.order_creation.buy_token,
+                        liquidity_order_trade.buy_token_price,
+                    )
+                })
+                .unzip();
+        let uniform_clearing_price_vec_length = self.tokens.len();
+        let mut tokens = self.tokens.clone();
```

Comment:
Not too long ago `drop_unnecessary_tokens_and_prices()` was introduced to make sure the settlement data has no unnecessary token or price data to save gas costs.
Judging by the test `settlement_reflects_different_price_for_normal_and_liquidity_order` duplicated tokens can appear in the final settlement data now.
I don't think this PR has to be updated to include this optimization but we should be aware of it and maybe add a comment or open an issue to track that.

Reply:
Thanks for your comments. They are all great.

Regarding this specific one: Yes, duplicated tokens can appear in the final settlement now. They are required to provide the different prices for the same token. 
Previously, we had only 1 price per token. But with this new PR, we want to settle liquidity orders at different prices. Hence, we no longer have 1 price per token, but can several prices. To provide these several prices, we need to provide the token as reference several times.

But for sure, I agree with you that `drop_unnecessary_tokens_and_prices` could double check that 
there are no double entries for the same token - the double entries will be added later. But let's do this in another PR, as even your original implementation does not do that - it only kick's out unnecessary ones.

Reply:
Ahh, I forgot about the 2 different prices for the same token.
Thanks for the explanation. 👍